Mathlib/Topology/Compactness/CompactlyCoherentSpace.lean

```diff
@@ -0,0 +1,336 @@
+  map from `CompactCoherentification X` to `CompactCoherentification Y`.
+
+## References
+* [J. Munkres, *Topology*][Munkres2000]
+* <https://en.wikipedia.org/wiki/Compactly_generated_space>
+-/
+
+noncomputable section
+
+open Set Set.Notation Topology
+
+/-! ### Compactly coherent spaces-/
+
+/-- A space is a compactly coherent space if the topology is generated by the compact sets. -/
+class CompactlyCoherentSpace  (X : Type*) [TopologicalSpace X] : Prop where
+  /-- A space is a compactly coherent space if the topology is generated by the compact sets.-/
+  isCoherentWith : IsCoherentWith (X := X) {K | IsCompact K}
+
+namespace CompactlyCoherentSpace
+
+universe u
+
+variable {X : Type u} [TopologicalSpace X]
+
+/-- A set `A` in a compactly coherent space is open iff for every compact set `K`, the intersection
+`K ∩ A` is open in `K`. -/
+lemma isOpen_iff [CompactlyCoherentSpace X] (A : Set X) :
+    IsOpen A ↔ ∀ K, IsCompact K → IsOpen (K ↓∩ A) :=
+  IsCoherentWith.isOpen_iff isCoherentWith
+
+/-- A set `A` in a compactly coherent space is closed iff for every compact set `K`, the
+intersection `K ∩ A` is closed in `K`. -/
+lemma isClosed_iff [CompactlyCoherentSpace X] (A : Set X) :
+    IsClosed A ↔ ∀ K, IsCompact K → IsClosed (K ↓∩ A) :=
+  IsCoherentWith.isClosed_iff isCoherentWith
+
+/-- If every set `A` is open if for every compact `K` the intersection `K ∩ A` is open in `K`,
+then the space is a compactly coherent space. -/
+lemma of_isOpen (h : ∀ (A : Set X), (∀ K, IsCompact K → IsOpen (K ↓∩ A)) → IsOpen A) :
+    CompactlyCoherentSpace X where
+  isCoherentWith := {isOpen_of_forall_induced := h}
+
+/-- If every set `A` is closed if for every compact `K` the intersection `K ∩ A` is closed in `K`,
+then the space is a compactly coherent space. -/
+lemma of_isClosed (h : ∀ (A : Set X), (∀ K, IsCompact K → IsClosed (K ↓∩ A)) → IsClosed A) :
+    CompactlyCoherentSpace X where
+  isCoherentWith := IsCoherentWith.of_isClosed h
+
+/-- Every weakly locally compact space is a compactly coherent space. -/
+instance of_weaklyLocallyCompactSpace [WeaklyLocallyCompactSpace X] : CompactlyCoherentSpace X where
+  isCoherentWith := IsCoherentWith.isCompact_of_weaklyLocallyCompact
+
+/-- Every sequential space is a compactly coherent space. -/
+instance of_sequentialSpace [SequentialSpace X] : CompactlyCoherentSpace X where
+  isCoherentWith := IsCoherentWith.isCompact_of_seq
+
+/-- In a compactly coherent space `X`, a set `s` is open when `f ⁻¹' s` is open for every continuous
+map from a compact space. -/
+lemma preimage_isOpen [CompactlyCoherentSpace X] (s : Set X)
+    (hs : (∀ (K : Type u) [TopologicalSpace K],
+      [CompactSpace K] → ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s))) :
+    IsOpen s := by
+  rw [isOpen_iff]
+  intro K hK
+  let _ : CompactSpace (Elem K) := isCompact_iff_compactSpace.mp hK
+  exact hs (Elem K) Subtype.val continuous_subtype_val
```

Comment:
I would upgrade this to an iff:
```suggestion
lemma isOpen_iff_forall_compactSpace [CompactlyCoherentSpace X] (s : Set X) : 
    IsOpen s ↔ 
      ∀ (K : Type u) [TopologicalSpace K] [CompactSpace K], 
      ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s) := by
  refine ⟨fun hs _ _ _ _ hf ↦ hs.preimage hf, fun hs ↦ isOpen_iff _ |>.mpr ?_⟩
  intro K hK
  let _ : CompactSpace K := isCompact_iff_compactSpace.mp hK
  exact hs K Subtype.val continuous_subtype_val
```
(We don't usually spell out `Elem`, instead we let the coercion mechanism do its magic)